src/OrleansRuntime/Streams/PubSub/PubSubRendezvousGrain.cs

```diff
@@ -276,14 +277,23 @@ private void RemoveProducer(PubSubPublisherState producer)
             State.Producers.Remove(producer);
         }
 
-        public async Task UnregisterConsumer(GuidId subscriptionId, StreamId streamId)
+        public async Task UnregisterConsumer(GuidId subscriptionId, StreamId streamId, bool observerRemoved)
         {
             counterConsumersRemoved.Increment();
             if (State.Consumers.Any(c => c.IsFaulted && c.Equals(subscriptionId)))
                 throw new FaultedSubscriptionException(subscriptionId, streamId);
 
             try
             {
+                if (!observerRemoved)
+                {
+                    //remove observer from consumer extension if not removed yet
+                    //this would happen when subscription removed by calling StreamSubscriptionManager
+                    var tasks = State.Consumers.Where(c => c.Equals(subscriptionId))
```

Comment:
I'm thinking about dropping onSubscriptionRemove behavior support. I don't see many users will need it and it will require us to call the consumer grain from here , which may result in an long blocking call and time out UnregisterConsumer call